app/javascript/src/PaymentGateways/braintree/BraintreeForm.jsx

```diff
@@ -0,0 +1,149 @@
+import { CSRFToken } from 'utilities/CSRFToken'
+import validate from 'validate.js'
+import './styles.css'
+
+import type { Node } from 'react'
+import type { BraintreeFormProps, BillingAddressData } from 'PaymentGateways'
+
+const BraintreeForm = ({
+  braintreeClient,
+  billingAddress,
+  threeDSecureEnabled,
+  formActionPath,
+  countriesList,
+  selectedCountryCode
+}: BraintreeFormProps): Node => {
+  const formRef = useRef<HTMLFormElement | null>(null)
+  // eslint-disable-next-line flowtype/no-weak-types
+  const [hostedFieldsInstance, setHostedFieldsInstance] = useState<any>(null)
+  const [braintreeNonceValue, setBraintreeNonceValue] = useState('')
+  const [billingAddressData, setBillingAddressData] = useState<BillingAddressData>(billingAddress)
+  const [isCardValid, setIsCardValid] = useState(false)
+  const [isFormValid, setIsFormValid] = useState(false)
+  const [formErrors, setFormErrors] = useState(validate(formRef, validationConstraints))
+  const [cardError, setCardError] = useState(null)
+  const [isAwaiting, setIsAwaiting] = useState(false)
+
+  useEffect(() => {
+    const getHostedFieldsInstance = async () => {
+      const HFInstance = await createHostedFieldsInstance(hostedFields, braintreeClient, hostedFieldOptions, setIsCardValid, setCardError)
+      setHostedFieldsInstance(HFInstance)
+    }
+    getHostedFieldsInstance()
+  }, [])
+
+  useEffect(() => {
+    if (braintreeNonceValue && formRef.current) {
+      formRef.current.submit()
+    }
+  }, [braintreeNonceValue])
+
+  useEffect(() => {
+    const formValid = isCardValid && !formErrors && !isAwaiting
+    setIsFormValid(formValid)
+  }, [formErrors, isCardValid, isAwaiting])
+
+  const get3DSecureNonce = async (payload) => {
+    const threeDSecureInstance = await create3DSecureInstance(threeDSecure, braintreeClient)
+    const response = await veryfyCard(threeDSecureInstance, payload, billingAddressData)
+    const error = response.name === 'BraintreeError'
+      ? response.code === 'THREEDS_LOOKUP_VALIDATION_ERROR' ? response.details.originalError.details.originalError.error.message : response.message
+      : null
+    const nonce = response.nonce || null
+
+    return {
+      error,
+      nonce
+    }
+  }
+
+  const clearHostedFields = () => {
+    hostedFieldsInstance.clear('number')
+    hostedFieldsInstance.clear('cvv')
+    hostedFieldsInstance.clear('expirationDate')
+  }
+
+  const validateForm = (event: SyntheticEvent<HTMLFormElement>) => {
+    const validationErrors = validate(event.currentTarget, validationConstraints)
+    setFormErrors(validationErrors)
+  }
+
+  const handleCardError = (error: string) => {
+    setCardError(`Credit card errors found: ${error}. Please correct your CC data.`)
+    clearHostedFields()
+    setIsAwaiting(false)
+  }
+
+  const onSubmit = async (event: SyntheticEvent<HTMLInputElement>) => {
+    event.preventDefault()
+    if (hostedFieldsInstance) {
+      setIsAwaiting(true)
```

Comment:
Is `isAwaiting` to prevent user interactions when the page is "busy"? If so:
* I strongly suggest renaming it to `isLoading` as I think it's more conventional.
* I may be wrong but since the whole block is async, this won't do anything. The page will simply freeze until `onSubmit` returns. You could verify this by adding some logs: _onSubmit starts_, _onSubmit ends_.. and see if they are being printed all at once.

Reply:
I've checked and both logs prints at same time

Reply:
Oh, OK that's also important